### PR TITLE
Add recaptcha to donations form

### PR DESCRIPTION
**Pull request by ErikSchierboom**
- Use endless method in captcha
- Render recaptcha
- Fix recaptcha CSP warnings: https://developers.google.com/recaptcha/docs/faq

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -96,8 +98,8 @@ def csp_policy
       connect: ["'self'", websockets, spellchecker],
       img: %w['self' data: https://*],
       media: %w[*],
-      script: default + [stripe, spellchecker],
-      frame: [stripe],
+      script: default + [stripe, spellchecker, google_recaptcha, gstatic_recaptcha],
+      frame: [stripe, google_recaptcha, gstatic_recaptcha],
```
Comment (ErikSchierboom) [inline app/controllers/application_controller.rb]: See https://developers.google.com/recaptcha/docs/faq

File: app/helpers/react_components/donations/footer_form.rb [hunk 2]
```diff
@@ -12,6 +12,8 @@ def to_s
               }
             },
             user_signed_in: user_signed_in?,
+            captcha_required: !current_user&.github_auth?,
+            recaptcha_site_key: '6LfFYEUjAAAAAH9eRl1qeO2R9aXzdXGnAybe6ulM', # TODO: use secret
```
Comment (ErikSchierboom) [inline app/helpers/react_components/donations/footer_form.rb]: @iHiD We'll probably want to store this as a secret (I'll deprecate this key)

Comment (iHiD) [reply app/helpers/react_components/donations/footer_form.rb]: Yep - you ok to do it?

Comment (ErikSchierboom) [reply app/helpers/react_components/donations/footer_form.rb]: Sure

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -83,6 +83,8 @@ def set_request_context
   def csp_policy
     websockets = "ws://#{Rails.env.production? ? 'exercism.org' : 'local.exercism.io:3334'}"
     stripe = "https://js.stripe.com"
+    google_recaptcha = "https://www.google.com/recaptcha/"
+    gstatic_recaptcha = "https://www.gstatic.com/recaptcha/"
```
Comment (iHiD) [inline app/controllers/application_controller.rb]:
```suggestion
    captcha = %w(https://www.google.com/recaptcha/ https://www.gstatic.com/recaptcha/)
```

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (iHiD) [inline app/controllers/application_controller.rb]:
```suggestion
      script: default + [stripe, spellchecker, *captcha],
      frame: [stripe, *captcha],
```

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (iHiD) [reply app/controllers/application_controller.rb]: I wonder if we need to add hcaptcha here too.

File: app/helpers/view_components/site_footer.rb [hunk 4]
```diff
@@ -18,5 +15,26 @@ def to_s
...
+      parts << user_part
+
+      parts.join(':')
+    end
+
+    def user_part
+      return 0 unless user_signed_in?
+
+      current_user.captcha_required? ? 2 : 1
+    end
```
Comment (ErikSchierboom) [inline app/helpers/view_components/site_footer.rb:32]: This ensures that the cache includes whether or not the user is required for a captcha to be shown

File: app/helpers/view_components/site_footer.rb [hunk 5]
```diff
@@ -18,5 +15,26 @@ def to_s
         end
       end
     end
+
+    private
+    def cache_key
```
Comment (ErikSchierboom) [inline app/helpers/view_components/site_footer.rb:20]: I took the liberty to slightly rewrite this as adding another clause was unwieldy

**Pull request merged**